0005-kzen-two-party-ecdsa.md

```diff
@@ -0,0 +1,139 @@
+
+The protocol is roughly as follows:
+
+- P1 -> Hash(p1_public_key, schnorr_nizk(p1_public_key))
+- P2 -> p2_public_key, schnorr_nizk(p2_public_key)
+- P1 -> Opens commitment from first message
+- Both parties calculate `joint_public_key = p1_public_key + p2_public_key`.
+
+#### Implementation notes
+The implementation commits by sending two hashes, one for the key and one for the NIZK.
+
+The NIZK is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_dlog.rs
+
+### Proving Paillier modulus is well formed
+
+In order to use Paillier encryption P1 needs to generate a RSA type modulus `N = pq` where `p` and `q` are large primes.
+P1 needs to prove that she generated `N` properly (I actually don't know the details of what the risk is to P2 if she doesn't).
+
+The proof mentioned in the original paper is this one: https://eprint.iacr.org/2011/494.
+But in a more recent paper on multi-party ECDSA Lindell mentioned a more efficient protocol to make the proof: [https://eprint.iacr.org/2018/987.pdf](https://eprint.iacr.org/2018/987.pdf).
+See the section 6.2.3, which explains how to modify the proof in [https://eprint.iacr.org/2018/057.pdf](https://eprint.iacr.org/2018/057.pdf) to simply prove `N` is generated fairly.
+
+Roughly speaking, the strategy is to prove that `N` is *square free* and to show that `f(x) = x ^ N (mod N)` is a permutation which proves that `N = pq`.
+
+#### Implementation notes
+
+Since it's a public coin protocol, you can make an non-interactive version of the proof using the [Fiat-Shamir transformation](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic).
+They've implemented a non-interactive version of it here: https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/correct_key_ni.rs (which uses the interactive proof module but supplies the challenge as a hash of the initial commitment).
+
+### Proving a Paillier encryption is a private key of a EC public key
+
+In order to produce two party aggregated ECDSA signatures on a joint Secp256k1 public key we need one party (P1) their Paillier encrypted private key over to P2.
+But in order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof for this was developed in paper. It's a little tricky to get your head around.
+It works by the P2 doing some homomorphic operations on the encrypted private key and sending the result to P1.
+P1 then decrypts it and figures out the corresponding public key.
+
+The weird thing is that P1 doesn't actually use his knowledge of what the private key is in order to produce the proof that the cipher text doesn't contain the private key.
+He simply uses his ability to decrypt the Paillier private key!
```

Comment:
I am not sure if I get this.
If it is important, it would be good to elaborate on this a bit more.

Reply:
Heh that doesn't should be a "does". But yes it is hard to explain. I'll see what i can do.